From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Grant Oberhauser <oberhausergrant@gmail.com>
Date: Sun, 8 Mar 2026 19:04:17 -0700
Subject: [PATCH] Mica: Fix vanilla redstone evaluator for parallel execution


diff --git a/io/papermc/paper/redstone/RedstoneWireTurbo.java b/io/papermc/paper/redstone/RedstoneWireTurbo.java
index ff747a1ecdf3c888bca0d69de4f85dcd810b6139..1e5841a866a0430209b48b005a029e3f37410f7e 100644
--- a/io/papermc/paper/redstone/RedstoneWireTurbo.java
+++ b/io/papermc/paper/redstone/RedstoneWireTurbo.java
@@ -829,14 +829,14 @@ public final class RedstoneWireTurbo {
         j = getMaxCurrentStrength(upd, j);
         int l = 0;
 
-        wire.shouldSignal = false;
+        wire.shouldSignal.set(false);
         // Unfortunately, World.isBlockIndirectlyGettingPowered is complicated,
         // and I'm not ready to try to replicate even more functionality from
         // elsewhere in Minecraft into this accelerator.  So sadly, we must
         // suffer the performance hit of this very expensive call.  If there
         // is consistency to what this call returns, we may be able to cache it.
         final int k = worldIn.getBestNeighborSignal(upd.self);
-        wire.shouldSignal = true;
+        wire.shouldSignal.set(true);
 
         // The variable 'k' holds the maximum redstone power value of any adjacent blocks.
         // If 'k' has the highest level of all neighbors, then the power level of this
diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index 6d98bc37d88459d1e0a171b52bbff5810d775c38..f54b40dc2507c222b7fe2262eb8b998b73ca8565 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -64,7 +64,10 @@ public class RedStoneWireBlock extends Block {
     private final Function<BlockState, VoxelShape> shapes;
     private final BlockState crossState;
     private final RedstoneWireEvaluator evaluator = new DefaultRedstoneWireEvaluator(this);
-    public boolean shouldSignal = true;
+
+    // Mica start - The "should signal" flag needs to be thread safe for parallel redstone evaluation.
+    public final ThreadLocal<Boolean> shouldSignal = ThreadLocal.withInitial(() -> Boolean.TRUE);
+    // Mica end
 
     @Override
     public MapCodec<RedStoneWireBlock> codec() {
@@ -327,9 +330,9 @@ public class RedStoneWireBlock extends Block {
     }
 
     public int getBlockSignal(Level level, BlockPos pos) {
-        this.shouldSignal = false;
+        this.shouldSignal.set(false);
         int bestNeighborSignal = level.getBestNeighborSignal(pos);
-        this.shouldSignal = true;
+        this.shouldSignal.set(true);
         return bestNeighborSignal;
     }
 
@@ -421,12 +424,12 @@ public class RedStoneWireBlock extends Block {
 
     @Override
     protected int getDirectSignal(BlockState state, BlockGetter level, BlockPos pos, Direction side) {
-        return !this.shouldSignal ? 0 : state.getSignal(level, pos, side);
+        return !this.shouldSignal.get() ? 0 : state.getSignal(level, pos, side);
     }
 
     @Override
     protected int getSignal(BlockState state, BlockGetter level, BlockPos pos, Direction side) {
-        if (this.shouldSignal && side != Direction.DOWN) {
+        if (this.shouldSignal.get() && side != Direction.DOWN) {
             int powerValue = state.getValue(POWER);
             if (powerValue == 0) {
                 return 0;
@@ -458,7 +461,7 @@ public class RedStoneWireBlock extends Block {
 
     @Override
     protected boolean isSignalSource(BlockState state) {
-        return this.shouldSignal;
+        return this.shouldSignal.get();
     }
 
     public static int getColorForPower(int power) {
